### PR TITLE
Implement async() functionality backed by a pool

### PR DESCRIPTION
**Pull request by dartdart26**
Implement an std::async()-like functionality that is backed by a thread
pool. Return std::future objects with the result to users. Support any
callable object with any return type.

Provide unit tests that also serve as examples.

Based on Mastering the C++17 STL, page 204:
https://books.google.bg/books?id=zJlGDwAAQBAJ&pg=PA205&lpg=PA205#

### CODE REVIEW

File: util/include/thread_pool.hpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    std::queue<GenericTask> tasks;
+
+    // A mutex for the queue.
+    std::mutex mutex;
+
+    // A condition variable to signal the queue.
+    std::condition_variable cv;
+  };
+
+  // Each thread has its own queue - rationale is that doing so will lessen the contention compared to having a single
```
Comment (toly-kournik) [inline util/include/thread_pool.hpp:109]: On the other hand starvation is very probable.
Personally I'd implement a single queue

Comment (dartdart26) [reply util/include/thread_pool.hpp:109]: Good point. Do you mean a case where tasks have different complexities?

I think, in general, we can measure it. Also, we can have a policy to specify if there should be one queue only. Another option is to implement work stealing.

Please let me know what your thoughts are on above.

Comment (dartdart26) [reply util/include/thread_pool.hpp]: After an offline discussion, I've pushed a commit that implements by using a single queue.

**Comment by dartdart26** [comment]
Will add a future commit with argument support as in `std::async()`,  i.e. `pool.async(func, arg1, arg2...)`.

File: util/include/thread_pool.hpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    }
+  }
+
+ public:
+  // Executes the passed function (or any callable) in a pool thread. Returns a future to the result.
+  // Note: The returned future's destructor doesn't block if the future value hasn't been retrieved. This is in contrast
+  // to the futures returned by std::async that block.
+  template <class F>
+  auto async(F&& func) {
+    using ResultType = std::invoke_result_t<std::decay_t<F>>;
```
Comment (andrewjstone) [inline util/include/thread_pool.hpp:63]: I've tried, but still don't fully understand why `std::decay_t<F>` is needed here. Do function pointers not work without it? What callable needs to be decayed?

File: util/include/thread_pool.hpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    {
+      auto lock = std::lock_guard{task_queue_.mutex};
+      task_queue_.tasks.push(std::move(task));
+    }
+    task_queue_.cv.notify_one();
+    return future;
+  }
+
+ private:
+  using GenericTask = std::packaged_task<void()>;
```
Comment (andrewjstone) [inline util/include/thread_pool.hpp:76]: Does this type erasure make it possible to add tasks with different return types? If so, can we do that in the tests as well please?

File: util/include/thread_pool.hpp (hunk 2 above)
Comment (dartdart26) [reply util/include/thread_pool.hpp:63]: It's a bit tricky and not extremely practical: https://stackoverflow.com/questions/37834567/decay-types-before-passing-to-stdresult-of

File: util/include/thread_pool.hpp (hunk 3 above)
Comment (dartdart26) [reply util/include/thread_pool.hpp:76]: Very good point. I have such a test in my next commit.

File: util/include/thread_pool.hpp (hunk 2 above)
Comment (andrewjstone) [reply util/include/thread_pool.hpp:63]: Thanks!

**Comment by dartdart26** [comment]
@HristoStaykov Thanks for the offline comment about the `non_blocking_future_dtors` test not doing what it is supposed to. Fixed by making sure the future's dtor is called while the lambda is still running.

**Comment by HristoStaykov** [comment]
Great work, Peter! User friendly implementation and very thorough unit tests.

**Pull request merged**